### PR TITLE
Prevent legacy extensions from targeting Firefox 57 or higher

### LINKED ISSUE #5277: All legacy add-ons (non-webextensions and themes) need a strict max version of Firefox 56.*
Firefox 57 won't support legacy add-ons. Specifically:

* All extensions that aren’t WebExtensions.
* Complete themes.

All legacy add-ons will need to show up as incompatible with Firefox 57 and above. So, for legacy add-ons that don't already have strict compatibility set, we should set it to a `maxVersion` of `56.*`. This only applies to compatibility with desktop Firefox, since compatibility with Android will probably be handled differently.

### PR DESCRIPTION
**Pull request by diox**
Also impose `strict_compatibility=True` on them to make sure they are never considered compatible.

Fix #5277

### CODE REVIEW

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+INNER JOIN `versions` ON ( `files`.`version_id` = `versions`.`id` )
+INNER JOIN `addons` ON ( `versions`.`addon_id` = `addons`.`id` )
+SET `strict_compatibility` = True
+WHERE (`files`.`is_webextension` = False AND `addons`.`addontype_id` IN (1, 2));
+
+
+-- All legacy add-ons must have a max below Firefox 57.
+-- Start by recording the new AppVersions we're going to set for those add-ons.
+SET @firefox_56 = (SELECT `id` FROM `appversions` WHERE `application_id` = 1 AND `version` = '56');
+SET @firefox_for_android_56 = (SELECT `id` FROM `appversions` WHERE `application_id` = 61 AND `version` = '56');
```
Comment (wagnerand) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: This one and the one above return empty sets on prod. The reason is that there is no  version '56'. We should probably set it to '56.0'.

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+WHERE (`applications_versions`.`application_id` = 61
+    AND `files`.`is_webextension` = False
+    AND `appversions`.`version_int` >= 57000000000000);
+
+-- Do the same thing for the max version now. Here, we can use 56.* as the last version.
+UPDATE `applications_versions`
+INNER JOIN `appversions` ON ( `applications_versions`.`max` = `appversions`.`id` )
+INNER JOIN `versions` ON ( `applications_versions`.`version_id` = `versions`.`id`
+INNER JOIN `files` ON ( `versions`.`id` = `files`.`version_id` )
+SET `applications_versions`.`max` = @firefox_below_57
```
Comment (wagnerand) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: I don't see where this var is defined.

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+WHERE (`applications_versions`.`application_id` = 1
+    AND `files`.`is_webextension` = False
+    AND `appversions`.`version_int` >= 57000000000000);
+
+UPDATE `applications_versions`
+INNER JOIN `appversions` ON ( `applications_versions`.`max` = `appversions`.`id` )
+INNER JOIN `versions` ON ( `applications_versions`.`version_id` = `versions`.`id`
+INNER JOIN `files` ON ( `versions`.`id` = `files`.`version_id` )
+SET `applications_versions`.`max` = @firefox_for_android_below_57
```
Comment (wagnerand) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: Same here, did you mean `@firefox_for_android_56_and_higher`?

Comment (diox) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: Yeah, I've been messing with that migration for too long... I'll fix it and retest it locally more thoroughly, thanks for catching that.

**Comment by wagnerand** [comment]
As discussed on IRC, jetpack add-ons will have a maxVersion of `*` by default. I am not sure, whether that should have an impact though.

File: src/olympia/devhub/forms.py [hunk 4]
```diff
@@ -432,14 +433,25 @@ class Meta:
...
             app = self.data[self.add_prefix('application')]
         self.app = amo.APPS_ALL[int(app)]
         qs = AppVersion.objects.filter(application=app).order_by('version_int')
+
+        # Legacy extensions can't set compatibility higher than 56.* for
+        # Firefox and Firefox for Android.
+        if (self.app in (amo.FIREFOX, amo.ANDROID) and
+                not version.is_webextension and
+                version.addon.type not in amo.NO_COMPAT):
+            qs = qs.filter(version_int__lt=57000000000000)
```
Comment (EnTeQuAk) [inline src/olympia/devhub/forms.py]: current nightly `57.0a1` has a `version_int` of `57000000001100` so this should work perfectly.

File: src/olympia/devhub/forms.py [hunk 5]
```diff
@@ -480,6 +495,27 @@ def clean(self):
...
                 ugettext('Need at least one compatible application.'))
 
+    # The 2 methods below, forms() and get_form_kwargs(), are lifted from
+    # Django 1.9, because we need form_kwargs to work.
+    @cached_property
+    def forms(self):
+        """
+        Instantiate forms at first property access.
+        """
+        # DoS protection is included in total_form_count()
```
Comment (EnTeQuAk) [inline src/olympia/devhub/forms.py]: reading that scares me tbh

File: src/olympia/devhub/tasks.py [hunk 6]
```diff
@@ -265,6 +268,19 @@ def handle_file_validation_result(results, file_id, *args):
     return FileValidation.from_json(file_, results)
 
 
+def insert_error_message(results, msg, msg_id):
```
Comment (EnTeQuAk) [inline src/olympia/devhub/tasks.py]: can we make the `type` configurable too? That'll allow other places to use that helper eventually.

Comment (EnTeQuAk) [inline src/olympia/devhub/tasks.py]: There's one other place where `compatibility_type` is set too. `description` is unused though

File: src/olympia/devhub/tests/test_tasks.py [hunk 7]
```diff
@@ -803,6 +812,79 @@ def test_restrict_firefox_53_alpha(self):
...
+
+        assert upload.processed_validation['errors'] == 0
+        assert upload.processed_validation['messages'] == []
+        assert upload.valid
+
+    def test_submit_dictionary_upgrade_targeting_firefox_57(self):
+        # Should not error: non-extensions types are not affected by the
+        # restriction, even if they target 57.
+        # FIXME
+        pass
```
Comment (EnTeQuAk) [inline src/olympia/devhub/tests/test_tasks.py]: that obviously needs to be fixed 😇

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 8]
```diff
@@ -0,0 +1,54 @@
+-- Enable strict compatibility for all extensions/themes that are not webextensions.
+UPDATE FROM `files`
+INNER JOIN `versions` ON ( `files`.`version_id` = `versions`.`id` )
+INNER JOIN `addons` ON ( `versions`.`addon_id` = `addons`.`id` )
+SET `strict_compatibility` = True
```
Comment (EnTeQuAk) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: should use `TRUE` or `FALSE` instead

**Comment by EnTeQuAk** [comment]
Does all this need a blog post to annouce this to developers cc  @jvillalobos?

File: src/olympia/devhub/forms.py (hunk 5 above)
Comment (diox) [inline src/olympia/devhub/forms.py]: That's copy/pasted from django :)

**Comment by diox** [comment]
> As discussed on IRC, jetpack add-ons will have a maxVersion of `*` by default. I am not sure, whether that should have an impact though.

My plan to address this is to allow legacy extensions to `*` without throwing an error, and instead silently rewrite `*` to `56.*` at extracting time, before creating the `File` and `Version` instances (only in the `RDFExtractor` that handles legacy extensions)

**Comment by wagnerand** [comment]
@jvillalobos should SDK extensions with *maxVersion* `*` be accepted and silently rewritten, as outlined in https://github.com/mozilla/addons-server/pull/6107#issuecomment-320232807? For context, XUL add-ons are not.

**Comment by diox** [comment]
> @jvillalobos should SDK extensions with maxVersion * be accepted and silently rewritten, as outlined in #6107 (comment)? For context, XUL add-ons are not.

I don't want to make a special case just for SDK, my proposal was to make a special case for `*`, *because* SDK defaults to `maxVersion` `*`. The question should rather be:

Should legacy extensions (SDK or XUL) with `maxVersion` set to `*` be accepted and silently rewritten as `56.*`, as outlined in #6107 (comment), or should it result in the same error as any version >= 57 ? If it should result in the same error, do we want a more generic error message than the one you provided, since it mentions only the `install.rdf` `maxVersion` property and not `engines` from `manifest.json`.

**Comment by jvillalobos** [comment]
> Does all this need a blog post

I don't think this particular issue warrants a blog post. But maybe it's worth writing something up for all the compatibility changes coming up. I'll look into it next week.

> Should legacy extensions (SDK or XUL) with maxVersion set to * be accepted and silently rewritten as 56.*

Rewriting sounds good to me.

File: src/olympia/devhub/tasks.py [hunk 9]
```diff
@@ -265,6 +268,22 @@ def handle_file_validation_result(results, file_id, *args):
...
+        'tier': 1,
+        'type': type_,
+        'id': ['validation', 'messages', msg_id],
+        'message': message,
+        'description': [],
+        'compatibility_type': compatibility_type,
+    })
+    # Need to increment 'errors' or 'warnings' count, so add an extra 's' after
+    # the type_ to increment the right entry.
+    results['{}s'.format(type_)] += 1
```
Comment (EnTeQuAk) [inline src/olympia/devhub/tasks.py]: I probably would have used a mapping for the `error` -> `errors` stuff but that works too 😆

Comment (EnTeQuAk) [inline src/olympia/devhub/tasks.py]: Thanks for updating this though!

File: src/olympia/files/tests/test_models.py [hunk 10]
```diff
@@ -420,11 +421,39 @@ def test_parse_permissions(self):
...
+                    application=amo.FIREFOX.id, version='38.0a1'),
+                AppVersion.objects.get(
+                    application=amo.FIREFOX.id, version='*')),
+
+            Extractor.App(
+                amo.ANDROID, amo.ANDROID.id,
+                AppVersion.objects.get(
+                    application=amo.ANDROID.id, version='38.0a1'),
+                AppVersion.objects.get(
+                    application=amo.ANDROID.id, version='*'))]
```
Comment (EnTeQuAk) [inline src/olympia/files/tests/test_models.py]: I might be blind but according to the comment this should be `version='56.*'` right?

File: src/olympia/files/utils.py [hunk 11]
```diff
@@ -263,10 +268,16 @@ def apps(self):
...
             seen_apps.add(app.id)
             try:
+                min_appver_text = self.find('minVersion', ctx)
+                max_appver_text = self.find('maxVersion', ctx)
+
+                if (app.id in (amo.FIREFOX.id, amo.ANDROID.id) and
+                        max_appver_text == '*'):
+                    # Rewrite '*' as '56.*' in legacy extensions, since they
+                    # are not compatible with higher versions.
+                    max_appver_text = '*'
```
Comment (EnTeQuAk) [inline src/olympia/files/utils.py]: This doesn't match the comment.

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    -- Start by recording the new AppVersions we're going to set for those add-ons.
+    SET @firefox_56 = (SELECT `id` FROM `appversions` WHERE `application_id` = 1 AND `version` = '56.0');
+    SET @firefox_for_android_56 = (SELECT `id` FROM `appversions` WHERE `application_id` = 61 AND `version` = '56.0');
+    SET @firefox_56_and_higher = (SELECT `id` FROM `appversions` WHERE `application_id` = 1 AND `version` = '56.*');
+    SET @firefox_for_android_56_and_higher = (SELECT `id` FROM `appversions` WHERE `application_id` = 61 AND `version` = '56.*');
+
+    -- Update ApplicationsVersions to set the new minimum version, if we're dealing with compatibility info
+    -- targeting Firefox/Firefox for Android >= 57.0 and not being webextensions. There should not be many of
+    -- them, but it's technically possible, since we let developers pick even '*' as a min version.
+    IF (@firefox_56 IS NOT NULL) THEN
```
Comment (EnTeQuAk) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: We should make sure they exist on -dev and -stage though before merging to test this migration properly

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        UPDATE `applications_versions`
+        INNER JOIN `appversions` ON ( `applications_versions`.`min` = `appversions`.`id` )
+        INNER JOIN `versions` ON ( `applications_versions`.`version_id` = `versions`.`id` )
+        INNER JOIN `files` ON ( `versions`.`id` = `files`.`version_id` )
+        SET `applications_versions`.`min` = @firefox_56
+        WHERE (`applications_versions`.`application_id` = 1
+            AND `files`.`is_webextension` = False
```
Comment (EnTeQuAk) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: `False` vs `FALSE` (the others below too)

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql (hunk 12 above)
Comment (diox) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: Double-checked: the 4 appversions do exist on dev/stage.

File: src/olympia/files/tests/test_models.py (hunk 10 above)
Comment (diox) [inline src/olympia/files/tests/test_models.py]: Oops, fixed.

File: src/olympia/files/utils.py (hunk 11 above)
Comment (diox) [inline src/olympia/files/utils.py]: Oops, fixed, like above: the comment was right, the code wasn't.

File: src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql (hunk 13 above)
Comment (diox) [inline src/olympia/migrations/962-change-legacy-addons-compatibility-for-57.sql]: done

**Pull request merged**